File: csharp/dnd-character/Iteration2/DndCharacter.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DndCharacter
{
    private DndCharacter() { }
    
    public int Strength { get; private set; }
    public int Dexterity { get; private set; }
    public int Constitution { get; private set; }
    public int Intelligence { get; private set; }
    public int Wisdom { get; private set; }
    public int Charisma { get; private set; }
    
    public int Hitpoints => 10 + Modifier(Constitution);

    public static int Modifier(int score) =>
        (int)Math.Floor((score - 10) / 2.0);
    
    public static int Ability() =>
        SixSidedDie
            .GetDiceRolls(4)
            .OrderByDescending(score => score)
            .Take(3)
            .Sum();

    public static DndCharacter Generate() =>
        new DndCharacter()
        {
            Strength = Ability(),
            Dexterity = Ability(),
            Constitution = Ability(),
            Intelligence = Ability(),
            Wisdom = Ability(),
            Charisma = Ability()
        };
}

public static class SixSidedDie
{
    private static Random _random = new Random();
    
    public static IEnumerable<int> GetDiceRolls(int number) =>
        GetDiceRolls().Take(number);
    
    public static IEnumerable<int> GetDiceRolls()
    {
        while (true)
            yield return Roll();
    }

    public static int Roll() => _random.Next(1, 7);
}

File: zig/dnd-character/Iteration2/dnd_character.zig
const std = @import("std");
const random = std.crypto.random;

pub fn modifier(score: i8) i8 {
    return @divFloor((score - 10), 2);
}

pub fn ability() i8 {
    var score: i8 = 0;
    var min: i8 = 6;

    inline for (0..4) |_| { 
        const val = random.intRangeAtMost(i8, 1, 6);
        score += val;
        min = @min(min, val);
    }

    score -= min;

    return score;
}

pub const Character = struct {
    strength: i8,
    dexterity: i8,
    constitution: i8,
    intelligence: i8,
    wisdom: i8,
    charisma: i8,
    hitpoints: i8,

    pub fn init() Character {
        var char: Character = undefined;

        inline for(@typeInfo(Character).Struct.fields) |field| {
            @field(char, field.name) = ability();
        }
            
        char.hitpoints = 10 + modifier(char.constitution);
        
        return char;
    }
};
